lara-typescript/src/interactive-api-client/hooks.ts

```diff
@@ -0,0 +1,86 @@
+import { useEffect, useState } from "react";
+import { IInitInteractive } from "./types";
+import * as client from "./api";
+
+export const useInteractiveState = <InteractiveState>() => {
+  const [ interactiveState, setInteractiveState ] = useState<InteractiveState | null>(client.getInteractiveState<InteractiveState>());
```

Comment:
I'm not sure if this is needed, but it might be. 

Since the client is actually maintaining the state there isn't really a need to store it in react's state.
So instead the line above could be:
`const [ , updateState ] = useState(0);`
So then the return from this hook would be `{ interactiveState: client.managedState.interactiveState, ...}`
The default value of `0` is so it can be incremented each time a change comes in.
That approach was taken from here: https://stackoverflow.com/a/55862077

Since React doesn't seem to do a deep compare of the state, as I noted below, calling setInteractiveState might not trigger the expected re-render.

But at the same time storing the actual state variable would might make it easier to use the React developer tools to see what is going on. But I think then there'd have to be a least a shallow copy of the interactiveState made on each update.

Reply:
Hm, so the counter approach would force re-render each time someone calls `setInteractiveState`. I can see two possible issues then. If we call `client.setInteractiveState` and return `client.managedState.interactiveState`, we implicitly assume that `client.setInteractiveState` **immediately** updates `client.managedState.interactiveState`. Of course, that's how things work in this particular case, but generally, it might not always be true (if there's some asyc operation, etc.). But this is more of a theoretical issue.

 > But I think then there'd have to be a least a shallow copy of the interactiveState made on each update.

Yes, I'm assuming that user of this hook would always do `Object.assign({}, interactiveState, {newProperty: 123})` or use something similar (jQuery extend, ImmutableJS, etc.). I'd say it's developer responsibility. Right now we're just consistent with `useState` behavior - if `useState` triggers re-render when the same object is passed, the developer is lucky. ;) If not, this is how React works (I believe it works that way and I always use Object.assign). 

Having said that, I don't think that forced re-renders would break things in practice or cause some performance issues in our case. But I think we'd add a pattern (forced updates) which is not fully consistent with react world/rules/hooks.

Reply:
Yes this makes sense.

Perhaps then, the setInteractiveState method should also accept a function (like setState) so the developer can do:
`setInteractiveState(prev => {...prev, newProperty: 123})`
Which would also be a more safe pattern if the setInteractiveState method is called inside of a callback. Since the interactiveState that is closed around would be frozen at the time of the last render.

Reply:
Thanks, actually I didn't know `setState` accepts it. I've added it in the last commit. 
btw, I would have to think more about it, but my first impression is that it's just for convenience. Not sure if I can come up with cases where it'd be safer than a regular update. I think that hook value is always equal to the value that is in functional component, as hook's state update means re-render.

Reply:
The `setState` function returned by `useState` is asynchronous in the same way that the class-based `setState` is, so if the state is updated between the time that the hook runs and the state is actually set, then the update would use the stale value rather than the most current value. I've seen examples with a button that increments a counter where if you press the button rapidly some clicks are "lost".

Reply:
I see. I'm wondering how to make it a problem in practice though.

Well, we can do:
```
setInteractiveState(Object.assign({}, intState, {foo: 1}));
setInteractiveState(Object.assign({}, intState, {bar: 2}));
```
And this will obviously fail. But the functional update I've added won't help. I guess React's implementation is smarter about this.

Also, I'm wondering what would be a more real-life scenario. Even if the interactive state is updated in some callback, I guess the previous state will be the same in the last render call and in the hook body (which is executed during... render). So, again, React's implementation must be smarter. 

The only case I can think of, where things can fail without obvious developer's fault, is when React is batching state updates. Then a few state updates can be executed at the same time. Each will have the same previous state, without new properties added by these calls. So, my fix won't help in this case either.

I'll think about some better implementation. I think I could pass `client.managedState.interactiveState` as a previous state to the update function. As `client.managedState.interactiveState` always gets updated synchronously, so it should be always more up to date than internal `setState` copy kept by React.

Thanks, I haven't thought about that before. :wink:

Reply:
Since we went deep here - if someone is interested, here's a test that actually shows the difference between non-functional and functional update:

https://github.com/concord-consortium/lara/blob/46e4baff5c2350ba7f6881713da6ba2de1416910/lara-typescript/src/interactive-api-client/hooks.spec.ts#L73-L93

And if I revert this commit/changes: https://github.com/concord-consortium/lara/pull/538/commits/9a376ed9a2011370eee060b4b2043278761c0472, this test of the functional updates fails.